File: pattern1/StrategyDesignPattern.cpp
#include<iostream>
 using namespace std;



 //clang++ -std=c++17 StrategyDesignPattern.cpp -o StrategyDesignPattern && ./StrategyDesignPattern
 
 
 class Talkable{
    public:
    virtual void talk()=0;
};
 class NormalTalk:public Talkable{
    public:
    void talk(){
        cout<<"normal Talk"<<endl;
    }
 };
  class StandardTalk:public Talkable{
    public:
    void talk(){
        cout<<"Standard Talk"<<endl;
    }
 };

  
 class Flyable{
    public:
    virtual void fly()=0;
};
 class NormalFly:public Flyable{
    public:
    void fly(){
        cout<<"normal fly"<<endl;
    }
 };
  class StandardFly:public Flyable{
    public:
    void fly(){
        cout<<"Standard fly"<<endl;
    }
 };

 class Robot{

    Talkable*talkbehaviour;
    Flyable*flybehaviour;
    public:
    Robot(Talkable* t,Flyable*f){
        this->talkbehaviour=t;
        this->flybehaviour=f;
    }
    void talk(){
        talkbehaviour->talk();
    }
    void fly(){
        flybehaviour->fly();
    }
 };
 class CompanionRobot :public Robot{
    public:
    CompanionRobot(Talkable*t,Flyable*f): Robot(t,f){}
 };


 int main(){
  Robot*r1= new CompanionRobot(new StandardTalk(),new NormalFly());
  r1->talk();
  r1->fly();
  return 0;
 }

File: pattern1/StrategyDesignPattern.java
package pattern1;

//defination
//Defines a family of algorithm(Talkable,Walkable ) , put them in separate classes so that they can be changed at runtime

interface Talkable{
    void talk();
}
class NormalTalk implements Talkable{
    public void talk(){
        System.out.println("Normal Talk");
    }
}
class StandardTalk implements Talkable{
    public void talk(){
        System.out.println("Standard Talk");
    }
}

interface Walkable{
    void walk();
}

class NormalWalk implements Walkable{
    public void walk(){
        System.out.println("Normal Walk");
    }
}
class StandardWalk implements Walkable{
    public void walk(){
        System.out.println("Standard Walk");
    }
}

 abstract class Robot{
    Talkable talkbehaviour;
    Walkable walkbehaviour;
    Robot(Talkable t, Walkable w){
        this.talkbehaviour=t;
        this.walkbehaviour=w;
    }
    void talk(){
        talkbehaviour.talk();
    }
    void walk(){
        walkbehaviour.walk();
    }
}

class CompanionRobot extends Robot{
    CompanionRobot(Talkable t, Walkable w){
        super(t,w);
    }
}



public class StrategyDesignPattern {
    public static void main(String[]args){
        System.out.println("hey");
        Robot r1= new CompanionRobot(new NormalTalk(), new StandardWalk());
        r1.walk();
        r1.talk();
    }
}
